### DISCUSSION TITLE
Exceptions on DBTreeview / Delphi 12

### DISCUSSION THREAD

**Discussion by mbaumsti** [project-jedi/jvcl#173] [Q&A]
Hello. The DBTreeView component doesn't work well with styles in Delphi 12, as it results in Access Violation exceptions when closing or deactivating the DataSource. When TStyleManager is not used, there are no issues.

**Comment by obones** [comment]
Can you provide the following elements?

Did it work in a previous version of Delphi? If yes, which one?
What is the call stack of the exception?
Are you using the latest Git sources unmodified?

**Reply by mbaumsti** [reply]
Hello. 
1) Here is the create evenment of my form 

Procedure TFormGesWeb.SelAppColors;
Begin
    if lowercase(extractfilename(Application.Exename)) = 'gesweb2.exe' then
    Begin
        ColFond := TStyleManager.ActiveStyle.GetStyleColor(scTreeView);
        ColSel := TStyleManager.ActiveStyle.GetSystemColor(clHighLight);
        ColFoc := TStyleManager.ActiveStyle.GetStyleColor(scButtonFocused);

        ColTxFondVert := DecalLumCol(TColor(RGB(0, 70, 0)), ColFond);
        ColTxSelVert := DecalLumCol(TColor(RGB(0, 70, 0)), ColSel);
        ColTxFocVert := DecalLumCol(TColor(RGB(0, 70, 0)), ColFoc);
    End;
End;

Procedure TFormGesWeb.ChangeApplicationAltColor;
var
    Col:     TColor;
    H, S, L: Single;

Begin
    if lowercase(extractfilename(Application.Exename)) = 'gesweb2.exe' then
    Begin

//        for var j := 0 to Application.ComponentCount - 1 do
//            If Application.Components[j] is TForm Then
//                CHangeAlternateColor(TForm(Application.Components[j]));

        SelAppColors;
    End;
End;

procedure TFormGesWeb.JvFormStorage1StoredValues0Restore(Sender: TJvStoredValue; var AValue: Variant);
begin
    if lowercase(extractfilename(Application.Exename)) = 'gesweb2.exe' then
    Begin
        Style := AValue;
        cbxVclStyles.ItemIndex := cbxVclStyles.Items.IndexOf(AValue);
    End;
end;

procedure TFormGesWeb.JvFormStorage1StoredValues0Save(Sender: TJvStoredValue; var AValue: Variant);
begin
    if lowercase(extractfilename(Application.Exename)) = 'gesweb2.exe' then
    Begin
        AValue := TStyleManager.ActiveStyle.Name;
    End;
end;

procedure TFormGesWeb.cbxVclStylesChange(Sender: TObject);
begin
//exit;
    if lowercase(extractfilename(Application.Exename)) = 'gesweb2.exe' then
    Begin
        TStyleManager.SetStyle(cbxVclStyles.Text);
        ChangeApplicationAltColor;
    End;
end;

Procedure TFormGesWeb.FormCreate(Sender: TObject);
var
    StyleName: String;
    idx:       Integer;
Begin
    //Application.OnException:= MyExceptionHandler;
    MajAuto := false;
    OpenAppDatabase(EacDatabaseGesWeb2, Self, '');

    TableArbo.Open;
    TableArboTrav.Open;
// exit;
    if lowercase(extractfilename(Application.Exename)) = 'gesweb2.exe' then
    Begin

        for StyleName in TStyleManager.StyleNames do
            cbxVclStyles.Items.add(StyleName);

        cbxVclStyles.ItemIndex := cbxVclStyles.Items.IndexOf(TStyleManager.ActiveStyle.Name);

        Style := RzRegIniFile1.ReadString('Prop', 'Style', '');
        if Style <> '' then
        Begin
            idx := cbxVclStyles.Items.IndexOf(Style);
            If idx <> -1 then
            Begin
                cbxVclStyles.ItemIndex := idx;
                TStyleManager.SetStyle(Style);
            End;
        end;

        CHangeAlternateColor(FormGesWeb);
        SelAppColors;
    End;
End;


2) Yes. It work in delphi 11.3


3) Here is the call stack 
:75e5f932 KERNELBASE.RaiseException + 0x62
:00a8bc48 NotifyNonDelphiException + $1C
:77dd8a84 ; 
:77dc50d6 ntdll.KiUserExceptionDispatcher + 0x26
:00bf3109 TControl.WndProc + $2C1
:00bf8750 TWinControl.WndProc + $6AC
:00c53153 TCustomTreeView.WndProc + $87
:00bf2d3f TControl.Perform + $27
:00bf88fc DoControlMsg + $28
:00bf8750 TWinControl.WndProc + $6AC
:00bf7c13 TWinControl.MainWndProc + $2F
:00b5d906 StdWndProc + $16
:76b30eab USER32.AddClipboardFormatListener + 0x4b
:76b27e5a ; C:\WINDOWS\SysWOW64\USER32.dll
:76b27239 ; C:\WINDOWS\SysWOW64\USER32.dll
:76b2587f USER32.SendMessageW + 0x6f
:71b29887 ; C:\WINDOWS\WinSxS\x86_microsoft.windows.common-controls_6595b64144ccf1df_6.0.19041.3636_none_a863d714867441db\COMCTL32.dll
:71afcfc7 ; C:\WINDOWS\WinSxS\x86_microsoft.windows.common-controls_6595b64144ccf1df_6.0.19041.3636_none_a863d714867441db\COMCTL32.dll
:71afd0d0 ; C:\WINDOWS\WinSxS\x86_microsoft.windows.common-controls_6595b64144ccf1df_6.0.19041.3636_none_a863d714867441db\COMCTL32.dll
:71afca10 ; C:\WINDOWS\WinSxS\x86_microsoft.windows.common-controls_6595b64144ccf1df_6.0.19041.3636_none_a863d714867441db\COMCTL32.dll
:71affbf6 ; C:\WINDOWS\WinSxS\x86_microsoft.windows.common-controls_6595b64144ccf1df_6.0.19041.3636_none_a863d714867441db\COMCTL32.dll
:76b30eab USER32.AddClipboardFormatListener + 0x4b
:76b27e5a ; C:\WINDOWS\SysWOW64\USER32.dll
:76b27537 ; C:\WINDOWS\SysWOW64\USER32.dll
:76b256db USER32.CallWindowProcW + 0x1b
:00bf88a6 TWinControl.DefaultHandler + $136
:00bf8750 TWinControl.WndProc + $6AC
:00c53153 TCustomTreeView.WndProc + $87
:00bf7c13 TWinControl.MainWndProc + $2F
:00b5d906 StdWndProc + $16
:76b30eab USER32.AddClipboardFormatListener + 0x4b
:76b27e5a ; C:\WINDOWS\SysWOW64\USER32.dll
:76b27239 ; C:\WINDOWS\SysWOW64\USER32.dll
:76b2587f USER32.SendMessageW + 0x6f
:00c4f6b2 TTreeNodes.Clear + $6A
:00c51893 TCustomTreeView.DestroyWnd + $1A7
JvDBTreeView.TJvCustomDBTreeView.DestroyWnd
:00bf75f7 TWinControl.DestroyHandle + $6B
:00bf7654 TWinControl.CMRecreateWnd + $3C
:00bf3109 TControl.WndProc + $2C1
:00bf8750 TWinControl.WndProc + $6AC
:00cf3300 TCustomForm.WndProc + $92C
:00bf2d3f TControl.Perform + $27
:00cf30b7 TCustomForm.WndProc + $6E3
:00bf7c13 TWinControl.MainWndProc + $2F
:00b5d906 StdWndProc + $16
:76b30eab USER32.AddClipboardFormatListener + 0x4b
:76b27e5a ; C:\WINDOWS\SysWOW64\USER32.dll
:76b25bca ; C:\WINDOWS\SysWOW64\USER32.dll
:76b25990 USER32.DispatchMessageW + 0x10

4) I downloaded zip files yesterday in this repository (no getit package found in delphi 12 ?)

thanks

**Comment by mbaumsti** [comment]
I tested close the table before change the style : 
procedure TFormGesWeb.cbxVclStylesChange(Sender: TObject);
begin
//exit;
    if lowercase(extractfilename(Application.Exename)) = 'gesweb2.exe' then
    Begin
TableArbo.Close;
        TStyleManager.SetStyle(cbxVclStyles.Text);
        ChangeApplicationAltColor;
TableArbo.Open;

    End;
end;

But I give the same error and compilator stop on the end this procedure : 
procedure TJvCustomDBTreeView.LinkActive(Value: Boolean);

  function AllFieldsValid: Boolean;
  begin
    Result := False;
    if ValidDataSet then
    begin
      if (FMasterField = '') or (FDataLink.DataSet.FindField(FMasterField) = nil) then
      begin
        Warning(RsMasterFieldEmpty);
        Exit;
      end;
      if (FDetailField = '') or (FDataLink.DataSet.FindField(FDetailField) = nil) then
      begin
        Warning(RsDetailFieldEmpty);
        Exit;
      end;
      if (FItemField = '') or (FDataLink.DataSet.FindField(FItemField) = nil) then
      begin
        Warning(RsItemFieldEmpty);
        Exit;
      end;
     { if (FDataLink.DataSet.FindField(FMasterField).DataType <> FDataLink.DataSet.FindField(FDetailField).DataType) then
       begin
        Warning(RsMasterDetailFieldError);
        Exit;
      end; }
      if (FDataLink.DataSet.FindField(FItemField).DataType in
        [ftBytes, ftVarBytes, ftBlob, ftGraphic, ftFmtMemo, ftParadoxOle, ftDBaseOle, ftTypedBinary]) then
      begin
        Warning(RsItemFieldError);
        Exit;
      end;
      if (FIconField <> '') and not (FDataLink.DataSet.FindField(FIconField).DataType in
        [ftSmallInt, ftInteger, ftWord]) then
      begin
        Warning(RsIconFieldError);
        Exit;
      end;
    end;
    Result := True;
  end;
begin
   if not Value then
    HideEditor;
  if not AllFieldsValid then
    Exit;
  //if ( csDesigning in ComponentState ) then Exit;
  if ValidDataSet then
  begin
    RefreshChild(nil);
    OldRecCount := FDataLink.DataSet.RecordCount;
  end
  else
    if (FUpdateLock = 0)  then
      Items.Clear;
end;

**Comment by mbaumsti** [comment]


**Reply by ronaldhoek** [reply]
Translated

> I told the debugger to ignore these types of exceptions... it works but it's not super clean. I really need to find a way to choose between open and execsql in my component...

**Reply by mbaumsti** [reply]
Excuse me this message is not for this thread, it is an error

**Reply by ronaldhoek** [reply]
Can you delete it please ;)

**Comment by alan008** [comment]
Check this:
https://github.com/project-jedi/jvcl/commit/ca21593219ff0f9ea837d428164105992d02dbcc

More details here:
https://github.com/project-jedi/jvcl/discussions/176#discussioncomment-8182425

**Comment by mbaumsti** [comment]
Ok solution work
Thanks
